packages/medusa/src/strategies/cart-completion.ts

```diff
@@ -245,26 +258,90 @@ class CartCompletionStrategy extends AbstractCartCompletionStrategy {
     }
 
     const orderServiceTx = this.orderService_.withTransaction(manager)
+    const swapServiceTx = this.swapService_.withTransaction(manager)
+    const cartServiceTx = this.cartService_.withTransaction(manager)
 
-    const cart = await this.cartService_
-      .withTransaction(manager)
-      .retrieveWithTotals(id, {
-        relations: ["region", "payment", "payment_sessions"],
-      })
+    const cart = await cartServiceTx.retrieveWithTotals(id, {
+      relations: ["region", "payment", "payment_sessions"],
+    })
+
+    let allowBackorder = false
+    let swapId: string
+
+    if (cart.type === "swap") {
+      const swap = await swapServiceTx.retrieveByCartId(id)
+      allowBackorder = swap.allow_backorder
+      swapId = swap.id
+    }
+
+    if (!allowBackorder) {
+      const productVariantInventoryServiceTx =
+        this.productVariantInventoryService_.withTransaction(manager)
+
+      try {
+        await Promise.all(
+          cart.items.map(async (item) => {
+            if (item.variant_id) {
+              const inventoryConfirmed =
+                await productVariantInventoryServiceTx.confirmInventory(
+                  item.variant_id,
+                  item.quantity,
+                  { salesChannelId: cart.sales_channel_id }
+                )
+
+              if (!inventoryConfirmed) {
+                throw new MedusaError(
+                  MedusaError.Types.NOT_ALLOWED,
+                  `Variant with id: ${item.variant_id} does not have the required inventory`,
+                  MedusaError.Codes.INSUFFICIENT_INVENTORY
+                )
+              }
+
+              await productVariantInventoryServiceTx.reserveQuantity(
+                item.variant_id,
+                item.quantity,
+                {
+                  lineItemId: item.id,
+                  salesChannelId: cart.sales_channel_id,
+                }
+              )
+            }
+          })
+        )
```

Comment:
thought: this should probably go into its own step -- but let's not do anything, as we will refactor to the transaction orchestrator in any case.

Reply:
👍